engine/src/main/java/org/hibernate/search/engine/search/projection/dsl/MultiProjectionTypeReference.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.search.engine.search.projection.dsl;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Set;
+import java.util.SortedSet;
+
+import org.hibernate.search.engine.search.projection.dsl.impl.ArrayMultiProjectionTypeReference;
+import org.hibernate.search.engine.search.projection.dsl.impl.ListMultiProjectionTypeReference;
+import org.hibernate.search.engine.search.projection.dsl.impl.SetMultiProjectionTypeReference;
+import org.hibernate.search.engine.search.projection.dsl.impl.SortedSetMultiProjectionTypeReference;
+import org.hibernate.search.util.common.annotation.Incubating;
+
+/**
+ * Defines the projection collection type to use for a multi-valued fields.
+ * <p>
+ * Either one of the built-in type references, provided by this interface, can be used,
+ * or a custom one can be implemented for other collection types.
+ * @param <C> The type of the collection.
+ * @param <V> The type of the elements in the collection.
+ */
+@Incubating
+public interface MultiProjectionTypeReference<C, V> {
```

Comment:
I couldn't come up with a better name for this one ....  so any suggestions are more than welcome 😃

Reply:
Couldn't you just promote `ProjectionAccumulator` to API, and use `ProjectionAccumulator.Provider`? It's admittedly a complex interface, but it's what we use under the hood, and if you provide implementations for the main container types, users will only need to implement it very rarely.

Then you could rename `ProjectionAccumulator#single()` to `nullable()` and introduce `optional()`.

Well, you'll probably need to create a new interface, and juggle with inheritance and deprecations, but you see what I mean.

Reply:
> It's admittedly a complex interface

yeah ... that's why I was thinking that since we are using this index-based approach for transformation operations in the accumulator that most implementations would still be backed by a list as a temporary storage option and the only difference would be the `R finish(A accumulated)` that performs the final transformation. Hence, I thought extracting this method (as `convert()`) would work for the users with some custom collections and we'd just delegate to in the list-based accumulator.

but let me see how promoting the accumulator to the API would work out 😃

Reply:
Fair enough, either solution has its merits. I'll let you have a look and decide :)

> Hence, I thought extracting this method (as `convert()`) would work for the users with some custom collections and we'd just delegate to in the list-based accumulator.

Note you could also provide a `static <V, C> Provider<V, C> simple(Function<List<V>, C> converter)` "factory method" to allow simple list-based implementations.

Reply:
> Note you could also provide a `static <V, C> Provider<V, C> simple(Function<List<V>, C> converter)` "factory method" to allow simple list-based implementations.

https://github.com/hibernate/hibernate-search/pull/4334 :grinning: